src/gurt/misc.c

```diff
@@ -947,149 +951,478 @@ d_rank_range_list_free(d_rank_range_list_t *range_list)
+	rc        = -DER_SUCCESS;
+
+out:
+	pthread_rwlock_unlock(&d_env_lock);
 
-	*bool_val = (atoi(env_val) == 0 ? false : true);
+	return rc;
 }
 
 /**
- * get single character environment variable
+ * get single character environment variable.
  *
- * \param[in]           env     name of the environment variable
- * \param[in,out]       char_val returned value of the ENV. Will not change the original value
+ * \param[in,out]	char_val	returned value of the ENV. Will not change the original
+ *					value.
+ * \param[in]		name		name of the environment variable.
+ * \return				0 on success, a negative value on error.
  */
-void
-d_getenv_char(const char *env, char *char_val)
+int
+d_getenv_char(char *char_val, const char *name)
 {
-	char		*env_val;
+	char *env;
+	int   rc;
 
-	if (env == NULL || char_val == NULL)
-		return;
+	assert(name != NULL);
+	assert(char_val != NULL);
 
-	env_val = getenv(env);
-	if (!env_val)
-		return;
+	pthread_rwlock_rdlock(&d_env_lock);
 
-	if (!dis_single_char_str(env_val)) {
-		D_ERROR("ENV %s is not single character.\n", env_val);
-		return;
+	env = getenv(name);
+	if (env == NULL) {
+		rc = -DER_NONEXIST;
+		goto out;
 	}
-	*char_val = *env_val;
+
+	if (!dis_single_char_str(env)) {
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	*char_val = *env;
+	rc        = -DER_SUCCESS;
+
+out:
+	pthread_rwlock_unlock(&d_env_lock);
+
+	return rc;
 }
 
 /**
- * get an integer type environment variables
+ * get an unsigned integer type environment variables.
  *
- * \param[in]		env	name of the environment variable
- * \param[in,out]	int_val	returned value of the ENV. Will not change the original value if ENV
- *				is not set or set as a
- *				non-integer value.
+ * \param[in,out]	uint_val	returned value of the ENV. Will not change the original
+ *					value if ENV is not set or set as a non-integer value.
+ * \param[in]		name		name of the environment variable.
+ * \return				0 on success, a negative value on error.
  */
-void
-d_getenv_int(const char *env, unsigned *int_val)
+int
+d_getenv_uint(unsigned *uint_val, const char *name)
 {
-	char		*env_val;
-	unsigned	 value;
+	char         *env;
+	char         *endptr;
+	unsigned long val;
+	int           rc;
+
+	assert(name != NULL);
+	assert(uint_val != NULL);
+
+	pthread_rwlock_rdlock(&d_env_lock);
+	env = getenv(name);
+	if (env == NULL) {
+		rc = -DER_NONEXIST;
+		goto out;
+	}
 
-	if (env == NULL || int_val == NULL)
-		return;
+	if (!dis_unsigned_str(env)) {
+		rc = -DER_INVAL;
+		goto out;
+	}
 
-	env_val = getenv(env);
-	if (!env_val)
-		return;
+	errno = 0;
+	val   = strtoul(env, &endptr, 0);
+	if (errno != 0 || endptr == env || *endptr != '\0') {
+		rc = -DER_INVAL;
+		goto out;
+	}
 
-	if (!dis_integer_str(env_val)) {
-		D_ERROR("ENV %s is not integer.\n", env_val);
-		return;
+#if UINT_MAX != ULONG_MAX
+	assert(sizeof(unsigned) < sizeof(unsigned long));
+	if (val > UINT_MAX) {
+		rc = -DER_INVAL;
+		goto out;
 	}
+#endif
```

Comment:
[suggestion] i think for somehow dis_unsigned_str() is duplicated as strtoull()?
we could remove dis_unsigned_str() and just use strtoull()?

Reply:
The best solution from my point of view should be to remove all the unsigned and to replace them with uint32_t or uint64_t.  However, this is beyond the scope of this PR.
For this PR, I would prefer to keep these three functions as the `sizeof()` of the `unsigned` is platform dependent: I would like to avoid any unexpected overflow.  In a follow-up PR, I could replace the `d_getenv_uint()` with relevant `d_getenv_uintXXt()` and update the code accordingly.

Reply:
I am fine with this, i thought dis_unsigned_str() is unnecessary and might be removed.

Reply:
- [ ] Factorize unsigned fonctions.

Reply:
> [suggestion] i think for somehow dis_unsigned_str() is duplicated as strtoull()? we could remove dis_unsigned_str() and just use strtoull()?

Sorry, i did not well understood your question regarding the function `dis_unsigned_str()`.  At the beginning, I also wanted to remove it.  However, `strtoull()` is less restrictive and allow to have spaces before the first digit and even to have a negative integer.  To not change the behavior of the `d_getenv_uint()` functions, I decided to kept the `dis_unsigned_str()` function.

Reply:
The code of the getenv of unsigned integer function has been factorized in commit 193c24c415c6e4229be3760821b8e7c9adeb3416